Giving python wrapper almost same functionalities and interface as C code

JaledMC opened the pull request:
First of all, thank for this well maintained repo, and kudos for the YOLOv4 step and its paper. It was written in a very instructive way.

## Why this pull request
Many useful features implemented in darknet C code are not accessed by the python wrapper, and the ones that can be used haven't the same commands. At the end, this situation becomes in a switching between both languages, and some incompatibilities with python MLOps tools.

## Features

* Added **colors for bboxes**, one for each class.
![Bbox colors](https://i.imgur.com/AWBiuCx.png)

* Added **command parser**.  At the moment, parameters for python scripts are handcoded. With this parser, we can invoke darknet like the README documentation:
```
python darknet_video.py --data_file cfg/coco.data --config_file cfg/yolov4.cfg --weights yolov4.weights --thresh 0.7 --input test.mp4 
```

* Print **bbox coordinates** with `--ext_output`
```
python darknet_video.py --weights yolov4.weights --input test.mp4 --ext_output
```

* Use `--dont_show` to prevent inference window. For headless systems:
```
python darknet_video.py --weights yolov4.weights --input test.mp4 --dont_show
```

* darknet as a module. Modules are the normal way to work with python, because of that, `darknet.py` is preserved, (and, as before, it doesn't need opencv installation). Examples are splited into `darknet_video.py` and `darknet_images.py` (which need opencv).

* **Save result videofile** video with `--out_filename`
```
python darknet_video.py --weights yolov4.weights --input test.mp4 --out_filename res.avi
```

* `darknet_video.py` can proccess **net-videocam** streams like `http://IP:PORT/xxxx.mjpg`

* `darnet_images.py`, like C code, accepts **path to and image** or a **txt file** with a list of them. If a **path to a folder** is given, it reads all jpg, png and jpeg images in it. If there is no path, it **loops asking** for one:
```
python darknet_images.py --weights yolov4.weights 
```
or
```
python darknet_images.py --weights yolov4.weights --input dog.jpg
```

* **Pseudolabeling** can be used in same way as in the README:
```
python darknet_images.py --weights yolov4.weights --input images/images_list.txt --save_labels
```

* Code cleaned: removed global variables, used functions in a proper way, and more readable code for less comments.


## TODO
* Code is not tested in Windows. I understand that this is a must :red_circle: 
* `AVG_fps` print is not implement.
* Bbox text is not resized with the its bbox size.
* An input parameter for desired class colors could be useful?
* Save inference images command (like `--output_file` video option)?


## My sin
This pull request is an only big commit :shit: I know, this is not the way to do things, but an unexpected docker shutdown removed all. A backup of the files survived :muscle:, but not the git folder. Sorry.

Hope it helps to anyone. If changes in the code are needed, feel free to ask.

cenit commented:
Wonderful, thanks.
I think we should work on the python interface a little bit after the continuous integration infrastructure update is in place. It’s a long time I promised @alexeyab to unify also the Darknet library and its behavior between the cmake version and the make version. Have you had time to verify it’s working with both, already?

cenit commented:
Also, we should set some basic test to verify python on all architectures. I think it might be easily doable if you have some time to spend on this front. In this way we can prepare an infrastructure that tests all OSes even if you don’t have access to all of them

AlexeyAB commented on darknet.py:
```diff
@@ -215,313 +302,3 @@ def network_height(net):
-    data = arr.ctypes.data_as(POINTER(c_float))
-    im = IMAGE(net_width, net_height, c, data)
-
-    batch_dets = network_predict_batch(net, im, batch_size, pred_width,
-                                                pred_height, thresh, hier_thresh, None, 0, 0)
```
Can we add param `--batch 4` to use `network_predict_batch()` for batch-processing of video and images for `darknet_video.py` and `darknet_images.py`?

AlexeyAB commented on darknet_video.py:
```diff
@@ -1,115 +1,103 @@
-        darknet.copy_image_from_bytes(darknet_image,frame_resized.tobytes())
+    cap = cv2.VideoCapture(args.input)
+    video = set_video(args.input, args.out_filename, (width, height))
 
-        detections = darknet.detect_image(netMain, metaMain, darknet_image, thresh=0.25)
-        image = cvDrawBoxes(detections, frame_resized)
+    while cap.isOpened():
+        prev_time = time.time()
+        ret, frame = cap.read()
+        if not ret:
+            break
+        frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
+        frame_resized = cv2.resize(frame_rgb, (width, height),
+                                   interpolation=cv2.INTER_LINEAR)
+        darknet.copy_image_from_bytes(darknet_image, frame_resized.tobytes())
+        detections = darknet.detect_image(network, class_names, darknet_image, thresh=args.thresh)
+        image = darknet.draw_boxes(detections, frame_resized, class_colors)
         image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
-        print(1/(time.time()-prev_time))
-        cv2.imshow('Demo', image)
-        cv2.waitKey(3)
+        if args.out_filename is not None:
+            video.write(image)
+        fps = int(1/(time.time() - prev_time))
+        print("FPS: {}".format(fps))
+        darknet.print_detections(detections, args.ext_output)
+        if not args.dont_show:
+            cv2.imshow('Inference', image)
+            cv2.waitKey(fps)
     cap.release()
-    out.release()
+    video.release()
```
Can we use 3 separate CPU-threads to process asychronously?
1. CPU-thread: VideoCapture + cvtColor + resize + copy_image_from_bytes
2. CPU-thread: darknet.detect_image
3. CPU-thread: draw_boxes + cvtColor + video.write + imshow + waitKey

JaledMC replied:
Good idea. I have to check that *point.2 CPU-thread: darknet.detect_image*, runs on GPU when possible, and python doesn't force CPU usage because bad multithread implementation.

IMHO, *point 1* and *point 3* can run always on CPU, because people rarely use opencv compiled with cuda. What your opinion?

AlexeyAB replied:
Yes, use CPU for points 1 and 2.

Since darknet.detect_image isn't async function, so creaste separate CPU-thread for it too.

So there should be 3 CPU-threads.

As in the C++ example - there are
*  1 default CPU-thread for Rendering(Show, Save to file, Send to web-browser)
and I create 2 additional CPU-threads: 
2. Inference (detection)
3. and VideoCapture+resizing

JaledMC commented:
Hi @cenit and @AlexeyAB 

I am agree with the improvements raised by you. It will take me some days to finish all tasks, but I am happy to take on the challenge :)

TODO list by order:
- [ ] Add performBatchDetect()
- [ ] Add batch feature to both `darknet_images.py` and `darknet_video.py`
- [ ] Use 3 threads: one for processing, one for model detection and another for rendering
- [ ] Check cmake vs make behavior
- [ ] CI test for OS compatibility

@cenit , excuse my ignorance but, can you refer me an example/tutorial about the CI test, to understand better which parts I have to check? Maybe [travis.yml](https://github.com/AlexeyAB/darknet/blob/master/.travis.yml) or [appveyor.yml](https://github.com/AlexeyAB/darknet/blob/master/appveyor.yml) can give me some guide?

KyloEntro commented:
Hi,

To achieve the take "Use 3 Threads", I wonder if it is interesting to go through python threading or multiprocessing to make CPU threading in python because of GIL which can block others threads until current thread finish its job.

AlexeyAB commented:
@JaledMC Hi, 
Did you manage to implement this functionality or did you encounter any difficulties?

JaledMC commented:
> @JaledMC Hi,
> Did you manage to implement this functionality or did you encounter any difficulties?

Excuse me @AlexeyAB , I have to solved some issues with a last project sprint at my work (working Saturday and Sunday too). Today is the deploy, and I will have this weekend free to do the task we agree.

Sorry for my lack of communication.

JaledMC commented:
> Hi,
> 
> To achieve the take "Use 3 Threads", I wonder if it is interesting to go through python threading or multiprocessing to make CPU threading in python because of GIL which can block others threads until current thread finish its job.

Hi @KyloEntro . Good point. I have the same question. But after a short investigation, I think multithread could be more appropriate:
* Frame reading is an I/O task, usually managed by threads.
* If I am right, yolo model uses ctypes, and GPU, bypassing the GIL.
* Multiprocessing has the advantage of been thread-safe, but process can't share memory space in an easy way (global variables, for example). Because of that, implementation could be harder.

IMHO, multithreading is a better idea, but I am open to changes

AlexeyAB commented:
@JaledMC Hi,
Did you manage to implement multithreaded asynchronous execution?

readicculus commented:
I got something sort of working a while back.  I can try and look into it again and clean it up.  Essentially got multiple batch detectors running on different threads but i had some issues because the optimal worker count was like 2 or 4 or something and if I went higher it would take longer. 


https://github.com/readicculus/darknet/blob/dcdbfce714d445885be02990e20d762b0bc4c24b/pydn/inference/darknet.py#L529-L544

JaledMC commented:
Hi @AlexeyAB 

Yes. It took me more time than I expected, but I think that both multithread video inference, and batch inference example, are done.

Three threads (for inference, video capture, and drawing) [have been implemented](https://github.com/Sialitech/darknet/blob/d0b6bae7316b576a221c4b4c1a461862bdf9cf43/darknet_video.py#L132-L134), using Queue to communicate and share variables between them. All previous features are included. There is one "small" problem: If *Escape* is clicked, frame window closes (as expected), and all processes stop, but Ctrl+C have to be used to free the *Terminal* again.

Batch inference example is not only preserved, but cleaned in a more pythonic way. We can add this [batch_inference](https://github.com/Sialitech/darknet/blob/d0b6bae7316b576a221c4b4c1a461862bdf9cf43/darknet_images.py#L118-L134) to the [main loop](https://github.com/Sialitech/darknet/blob/d0b6bae7316b576a221c4b4c1a461862bdf9cf43/darknet_images.py#L178-L214), to use it as another feature within the Terminal main command, but I haven't seen this option on the C interface, and maybe you have other idea. PS: tested with 3 images, because my GPU has small memory.

On the other hand, I haven't found a good way to implement batch inference for video... I can understand how to store a couple of frames, process the inference, but later we have to show each frame, one by one, without fps bouncing. With small GPUs, with less than 15 fps, that's tricky. The faster inference is enough for a more complex and delayed pipeline? I don't know.

As always, sorry for my delay, and hope this step bring a complete python option closer.

JaledMC commented:
PS: you have done a lot of changes since last time (congratulations!). If closing this pull request and opening a new one is better for you, let me know. I have developed this code with your last commit.

AlexeyAB commented on darknet.py:
```diff
@@ -61,26 +41,144 @@ class DETECTION(Structure):
                 ("sort_class", c_int),
                 ("uc", POINTER(c_float)),
-                ("points", c_int)]
+                ("points", c_int),
+                ("embeddings", POINTER(c_float)),
+                ("embedding_size", c_int),
+                ("sim", c_float),
+                ("track_id", c_int)]
```
> PS: you have done a lot of changes since last time (congratulations!). If closing this pull request and opening a new one is better for you, let me know. I have developed this code with your last commit.

You have added all the necessary changes, so this should work successfully.

AlexeyAB commented:
@JaledMC Hi,
Great work!

> On the other hand, I haven't found a good way to implement batch inference for video... I can understand how to store a couple of frames, process the inference, but later we have to show each frame, one by one, without fps bouncing. With small GPUs, with less than 15 fps, that's tricky. The faster inference is enough for a more complex and delayed pipeline? I don't know.

Fps bouncing for batch-inference on video - this is ok.
You can add switchable option to sync output-frames with input frames. Or a simpler option, calculate the average waiting time between batches, and add `cv2.waitKey(avg_batch_waiting_time_ms / batch_size)` after `cv2.imshow`

----

> PS: you have done a lot of changes since last time (congratulations!). If closing this pull request and opening a new one is better for you, let me know. I have developed this code with your last commit.

You have added all the necessary changes, so this should work successfully.

----

You just can add another one feature for tracking objects: https://github.com/AlexeyAB/darknet/blob/0b36b29b6aa6eddc2134581dd8f666d5db4276a9/include/darknet.h#L1076

```python
set_track_id= lib.set_track_id
set_track_id.argtypes = [POINTER(DETECTION), c_int, c_float, c_float, c_float, c_int, c_int, c_int]
```


Such code (just in Python) should be added after nms: https://github.com/AlexeyAB/darknet/pull/5445/files#diff-937e97eecf03db4d5ee5dada4367bcc8R174
```cpp
if (l.embedding_size) set_track_id(local_dets, local_nboxes, thresh, l.sim_thresh, l.track_ciou_norm, l.track_history_size, l.dets_for_track, l.dets_for_show);
```

It seems it requires to get the last [yolo] layer `l.` to get parameters `.sim_thresh, l.track_ciou_norm, l.track_history_size, l.dets_for_track, l.dets_for_show`
Perhaps I should add C-function `cotrastive_track_t get cotrastive_track_values(net);` that returns 
```c
struct cotrastive_track_t {
 float sim_thresh;
 float track_ciou_norm;
 int track_history_size;
 int dets_for_track;
 int dets_for_show;
 int embedding_size;
};
```

----

So I will merge this PR. Fill free to open new PRs for new features.

minertom commented:
I am working on a project that utilizes code such as darknet_video.py. It works fine under linux but will not work under the windows installation. 

Please test under windows and you will find this to be true.

You get a "module 'darknet' has no attribute" kind of error.

Tom

AlexeyAB commented:
@minertom 

It works well on **Windows 7 x64, python 3.6.4**, CUDA 10.1, cuDNN 7.6.5, OpenCV 4.2.0

![image](https://user-images.githubusercontent.com/4096485/89585300-47139f00-d846-11ea-9c67-2c91bb0f83eb.png)

![image](https://user-images.githubusercontent.com/4096485/89585352-5eeb2300-d846-11ea-90ad-8abe7cdea77f.png)

minertom commented:
Thank you so very much for this helpful information.

I am almost there. I did the execution from the x64 directory and received
the same error message.

File "darknet_video.py", line 58, in YOLO
    netMain = darknet.load_net_custom(configPath.encode(
AttributeError: module 'darknet' has no attribute 'load_net_custom'

when I am in the x64 directory and I run python, then import darknet,
darknet is, in fact, imported. So, I don't know where the error is.
Can you make a suggestion?

Thank You
Tom


On Thu, Aug 6, 2020 at 2:41 PM Alexey <notifications@github.com> wrote:

> @minertom <https://github.com/minertom>
>
> It works well on *Windows 7 x64, python 3.6.4*, CUDA 10.1, cuDNN 7.6.5,
> OpenCV 4.2.0
>
> [image: image]
> <https://user-images.githubusercontent.com/4096485/89585300-47139f00-d846-11ea-9c67-2c91bb0f83eb.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/4096485/89585352-5eeb2300-d846-11ea-90ad-8abe7cdea77f.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AlexeyAB/darknet/pull/5445#issuecomment-670205941>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHGGHGBKKCELVJEV7FC6O3R7MPPDANCNFSM4MXIVZZQ>
> .
>

AlexeyAB commented:
Compile `yolo_cpp_dll.sln` by using MSVS (or compile `darklib` by using Cmake).

minertom commented:
Which version of MSVS? I have version 16.6.5 from 2019 and solution build fails.

Tom

AlexeyAB commented:
MSVS2015 (14.0.25...)

minertom commented:
AlexeyAB

I am very close to a solution. One more step.

I was able to compile yolo_cpp_dll.sln by using MSVS2017 and changing the
build tools version to 141 as opposed to 142. The build went fine. I also
did a fresh CUDA 10.0 install.

As you can see from the attached screenshot, the yolo_cpp_dll.dll is
clearly in the directory with a timestamp for the build which does reflect
when it was done. But the error claims that this file is missing and can
not be found.

FileNotFoundError: Could not find module
'C:\darknet-master\build\darknet\x64\yolo_cpp_dll.dll' (or one of its
dependencies). Try using the full path with constructor syntax.

Not sure how to correct for this. Yes, doing >python import darknet, in any
directory where I believe that it should work, fails with the same error.
Could this be due to some environmental variable not being set?

Notice, also, that even though I specified "darknet_video.py" on the
command line, the command was attempting to execute "darknet.py". I am sure
I could delete "darknet.py" from the directory but should that be happening?

Thank You for all of your help.
Tom Cipollone



<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>


On Thu, Aug 6, 2020 at 2:41 PM Alexey <notifications@github.com> wrote:

> @minertom <https://github.com/minertom>
>
> It works well on *Windows 7 x64, python 3.6.4*, CUDA 10.1, cuDNN 7.6.5,
> OpenCV 4.2.0
>
> [image: image]
> <https://user-images.githubusercontent.com/4096485/89585300-47139f00-d846-11ea-9c67-2c91bb0f83eb.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/4096485/89585352-5eeb2300-d846-11ea-90ad-8abe7cdea77f.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AlexeyAB/darknet/pull/5445#issuecomment-670205941>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHGGHGBKKCELVJEV7FC6O3R7MPPDANCNFSM4MXIVZZQ>
> .
>

TheAshwanik commented:
> AlexeyAB I am very close to a solution. One more step. I was able to compile yolo_cpp_dll.sln by using MSVS2017 and changing the build tools version to 141 as opposed to 142. The build went fine. I also did a fresh CUDA 10.0 install. As you can see from the attached screenshot, the yolo_cpp_dll.dll is clearly in the directory with a timestamp for the build which does reflect when it was done. But the error claims that this file is missing and can not be found. FileNotFoundError: Could not find module 'C:\darknet-master\build\darknet\x64\yolo_cpp_dll.dll' (or one of its dependencies). Try using the full path with constructor syntax. Not sure how to correct for this. Yes, doing >python import darknet, in any directory where I believe that it should work, fails with the same error. Could this be due to some environmental variable not being set? Notice, also, that even though I specified "darknet_video.py" on the command line, the command was attempting to execute "darknet.py". I am sure I could delete "darknet.py" from the directory but should that be happening? Thank You for all of your help. Tom Cipollone <http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail> Virus-free. [www.avg.com](http://www.avg.com) <http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail> <#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>
> […](#)
> On Thu, Aug 6, 2020 at 2:41 PM Alexey ***@***.***> wrote: @minertom <https://github.com/minertom> It works well on *Windows 7 x64, python 3.6.4*, CUDA 10.1, cuDNN 7.6.5, OpenCV 4.2.0 [image: image] <https://user-images.githubusercontent.com/4096485/89585300-47139f00-d846-11ea-9c67-2c91bb0f83eb.png> [image: image] <https://user-images.githubusercontent.com/4096485/89585352-5eeb2300-d846-11ea-90ad-8abe7cdea77f.png> — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#5445 (comment)](https://github.com/AlexeyAB/darknet/pull/5445#issuecomment-670205941)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADHGGHGBKKCELVJEV7FC6O3R7MPPDANCNFSM4MXIVZZQ> .

Did you find a solution for this?

minertom commented:
Ashwani,

Thank you for your reply.

Actually, I did manage to fix the problem but I am not really sure of what
fixed it.

I determined that it had something to do with how python was looking for
its modules. I had reinstalled darknet from the github a couple of times
and had gone through the visual studio builds. I added a pythonpath
environmental variable but that  by itself did not work. Finally, I did one
more install of darknet from the github and for some reason, it started
working. Again, I am not sure of why it is working but I am not arguing.

Regards
Tom


On Sun, Aug 23, 2020 at 4:49 PM Ashwani Kumar <notifications@github.com>
wrote:

> AlexeyAB I am very close to a solution. One more step. I was able to
> compile yolo_cpp_dll.sln by using MSVS2017 and changing the build tools
> version to 141 as opposed to 142. The build went fine. I also did a fresh
> CUDA 10.0 install. As you can see from the attached screenshot, the
> yolo_cpp_dll.dll is clearly in the directory with a timestamp for the build
> which does reflect when it was done. But the error claims that this file is
> missing and can not be found. FileNotFoundError: Could not find module
> 'C:\darknet-master\build\darknet\x64\yolo_cpp_dll.dll' (or one of its
> dependencies). Try using the full path with constructor syntax. Not sure
> how to correct for this. Yes, doing >python import darknet, in any
> directory where I believe that it should work, fails with the same error.
> Could this be due to some environmental variable not being set? Notice,
> also, that even though I specified "darknet_video.py" on the command line,
> the command was attempting to execute "darknet.py". I am sure I could
> delete "darknet.py" from the directory but should that be happening? Thank
> You for all of your help. Tom Cipollone
> http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail
> Virus-free. www.avg.com
> http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail
> <#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>
> … <#m_-6468284461028127359_>
> On Thu, Aug 6, 2020 at 2:41 PM Alexey *@*.***> wrote: @minertom
> <https://github.com/minertom> https://github.com/minertom It works well
> on *Windows 7 x64, python 3.6.4*, CUDA 10.1, cuDNN 7.6.5, OpenCV 4.2.0
> [image: image]
> https://user-images.githubusercontent.com/4096485/89585300-47139f00-d846-11ea-9c67-2c91bb0f83eb.png
> [image: image]
> https://user-images.githubusercontent.com/4096485/89585352-5eeb2300-d846-11ea-90ad-8abe7cdea77f.png
> — You are receiving this because you were mentioned. Reply to this email
> directly, view it on GitHub <#5445 (comment)
> <https://github.com/AlexeyAB/darknet/pull/5445#issuecomment-670205941>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/ADHGGHGBKKCELVJEV7FC6O3R7MPPDANCNFSM4MXIVZZQ
> .
>
> Did you find a solution for this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AlexeyAB/darknet/pull/5445#issuecomment-678841370>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHGGHFEYXR7LCJ33ZMKF7LSCGTIBANCNFSM4MXIVZZQ>
> .
>

TheAshwanik commented:
Sadly adding pythonpath still does not work for me.

MysticCoss commented:
@TheAshwanik i have googled for more than 2 days and i actually find that in python 3.8 there were some major changes that made the code broken (on finding the dll). So i used python 3.7 and even 2.7 and it works. 
The code should be fixed but i dont have enough knowledge, so if someone has solution pls share. Thanks

cenit commented:
```
try:
    import os
    os.add_dll_directory(os.path.join(os.environ['CUDA_PATH'], 'bin'))
except Exception:
    pass
```

Did you build darknet.dll with or without CUDA? If you built it with CUDA, can you try adding this code at the beginning of the script?
